SU2_CFD/src/solvers/CFEASolver.cpp

```diff
@@ -0,0 +1,3774 @@
+        int EL_KIND;
+        unsigned short nNodes;
+        GetElemKindAndNumNodes(geometry->elem[iElem]->GetVTK_Type(), EL_KIND, nNodes);
+
+        /*--- Each thread needs a dedicated element. ---*/
+        CElement* element = element_container[FEA_TERM][EL_KIND+thread*MAX_FE_KINDS];
+
+        /*--- For the number of nodes, we get the coordinates from the connectivity matrix ---*/
+        unsigned long indexNode[MAXNNODE];
+
+        for (iNode = 0; iNode < nNodes; iNode++) {
+
+          indexNode[iNode] = geometry->elem[iElem]->GetNode(iNode);
+
+          for (iDim = 0; iDim < nDim; iDim++) {
+            /*--- Compute current coordinate. ---*/
+            su2double val_Coord = Get_ValCoord(geometry, indexNode[iNode], iDim);
+            su2double val_Sol = nodes->GetSolution(indexNode[iNode],iDim) + val_Coord;
+
+            /*--- If pre-stretched the reference coordinate is stored in the nodes. ---*/
+            if (prestretch_fem)
+              val_Coord = nodes->GetPrestretch(indexNode[iNode],iDim);
+
+            /*--- Set coordinates. ---*/
+            element->SetCurr_Coord(iNode, iDim, val_Sol);
+            element->SetRef_Coord(iNode, iDim, val_Coord);
+          }
+        }
+
+        /*--- In topology mode determine the penalty to apply to the stiffness ---*/
+        su2double simp_penalty = 1.0;
+        if (topology_mode) {
+          su2double density = element_properties[iElem]->GetPhysicalDensity();
+          simp_penalty = simp_minstiff+(1.0-simp_minstiff)*pow(density,simp_exponent);
+        }
+
+        /*--- Set the properties of the element. ---*/
+        element->Set_ElProperties(element_properties[iElem]);
+
+        /*--- Compute the components of the Jacobian and the stress term for the material. ---*/
+        int NUM_TERM = thread*MAX_TERMS + element_properties[iElem]->GetMat_Mod();
+
+        numerics[NUM_TERM]->Compute_NodalStress_Term(element, config);
+
+        for (iNode = 0; iNode < nNodes; iNode++) {
+          auto Ta = element->Get_Kt_a(iNode);
+          for (iVar = 0; iVar < nVar; iVar++)
+            LinSysRes[indexNode[iNode]*nVar+iVar] -= simp_penalty*Ta[iVar];
+        }
+
+      } // end iElem loop
+
+    } // end color loop
+
+  } // end SU2_OMP_PARALLEL
+
+}
+
+void CFEASolver::Compute_NodalStress(CGeometry *geometry, CNumerics **numerics, CConfig *config) {
+
+  const bool prestretch_fem = config->GetPrestretch();
+
+  const bool topology_mode = config->GetTopology_Optimization();
+  const su2double simp_exponent = config->GetSIMP_Exponent();
+
+  const unsigned short nStress = (nDim == 2) ? 3 : 6;
+
+  /*--- Reduction variable to compute the maximum von Misses stress,
+   *    each thread uses the first element of each row, the rest of
+   *    the row is padding to prevent false sharing. ---*/
+  su2activematrix auxMaxVonMisses(omp_get_max_threads(), 64/sizeof(su2double));
+  auxMaxVonMisses.setConstant(0.0);
```

Comment:
This is probably the trickiest bit in this PR, reduction operations in general are tricky, I can try to clean it up a bit by using the OpenMP reduction clause but this "manual" way is future compatible with the AD type (OpenMP reductions are only defined for builtin types, they can be declared for other types but that sounds like witchcraft to me).